controllers/scenario/scenario_adapter.go

```diff
@@ -0,0 +1,122 @@
+		scenario:    scenario,
+		logger:      logger,
+		client:      client,
+		context:     context,
+	}
+}
+
+// EnsureCreatedScenarioIsValid is an operation that ensures created IntegrationTestScenario is valid
+// in case it is, set its owner reference
+func (a *Adapter) EnsureCreatedScenarioIsValid() (reconciler.OperationResult, error) {
+
+	// first check if application exists or not
+	if a.application == nil {
+		a.logger.Info("Application for scenario was not found.",
+			"Sceario.Name", a.scenario.Name)
+
+		patch := client.MergeFrom(a.scenario.DeepCopy())
+		SetScenarioIntegrationStatusAsInvalid(a.scenario, "Failed to get application for scenario.")
+		return reconciler.RequeueOnErrorOrStop(a.client.Status().Patch(a.context, a.scenario, patch))
+		// if application exists, check if scenario has ownerReference assigned
+	} else if a.scenario.OwnerReferences == nil {
+		patch := client.MergeFrom(a.scenario.DeepCopy())
+		err := ctrl.SetControllerReference(a.application, a.scenario, a.client.Scheme())
+		if err != nil {
+			a.logger.Error(err, "Error setting owner reference.",
+				"Application.Name", a.application.Name,
+				"Application.Namespace", a.application.Namespace)
+			patch := client.MergeFrom(a.scenario.DeepCopy())
+			SetScenarioIntegrationStatusAsInvalid(a.scenario, "Failed to get application for scenario.")
+			return reconciler.RequeueOnErrorOrStop(a.client.Status().Patch(a.context, a.scenario, patch))
+		}
+		err = a.client.Patch(a.context, a.scenario, patch)
+		if err != nil {
+			a.logger.Error(err, "Failed to update Scenario",
+				"Scenario.Name", a.scenario.Name)
+			return reconciler.RequeueWithError(err)
+		}
+		return reconciler.ContinueProcessing()
+		// app exists, owner reference is set, check if the scenario has envirnment defined
+	} else if reflect.ValueOf(a.scenario.Spec.Environment).IsZero() {
+		a.logger.Info("IntegrationTestScenario has no environment defined",
+			"Scenario.Name:", a.scenario.Name,
+			"Scenario.Namespace", a.scenario.Namespace)
+		return reconciler.ContinueProcessing()
+	} else {
+		//Same as function getEnvironmentFromIntegrationTestScenario - this could be later changed to call only that function
+		ITSEnv := &applicationapiv1alpha1.Environment{}
+
+		err := a.client.Get(a.context, types.NamespacedName{
+			Namespace: a.application.Namespace,
+			Name:      a.scenario.Spec.Environment.Name,
+		}, ITSEnv)
+
+		if err != nil {
+			a.logger.Info("Environment doesn't exist in same namespace as IntegrationTestScenario.",
+				"Scenario.Name:", a.scenario.Name,
+				"Scenario.Namespace:", a.scenario.Namespace,
+				"Environment.Name:", a.scenario.Spec.Environment.Name)
+			patch := client.MergeFrom(a.scenario.DeepCopy())
+			SetScenarioIntegrationStatusAsInvalid(a.scenario, "Environment "+a.scenario.Spec.Environment.Name+" is located in different namespace than scenario.")
+			return reconciler.RequeueOnErrorOrStop(a.client.Status().Patch(a.context, a.scenario, patch))
+		}
+
+	}
```

Comment:
I think it would be good to update the IntegrationTestScenario status condition to valid here - if all of the above tests passed and if it wasn't already valid.

Reply:
Added, now the logic corresponds to this:
Valid scenario must meet those conditions:
- Has ENV specified
- ENV is situated within scenario namespace
- Has App defined

As a bonus it has owner reference added.